diff -ru vncsnapshot-1.2a.orig/rfb.h vncsnapshot-1.2a/rfb.h
--- vncsnapshot-1.2a.orig/rfb.h	2012-03-26 04:53:00.695298214 -0400
+++ vncsnapshot-1.2a/rfb.h	2012-03-26 04:53:33.759874425 -0400
@@ -34,7 +34,7 @@
 
 // Define the CARD* types as used in X11/Xmd.h
 
-typedef unsigned long CARD32;
+typedef unsigned int CARD32;
 typedef unsigned short CARD16;
 typedef short INT16;
 typedef unsigned char  CARD8;
diff -ru vncsnapshot-1.2a.orig/rfbproto.c vncsnapshot-1.2a/rfbproto.c
--- vncsnapshot-1.2a.orig/rfbproto.c	2012-03-26 04:53:00.699298283 -0400
+++ vncsnapshot-1.2a/rfbproto.c	2012-03-26 04:54:05.856430213 -0400
@@ -335,7 +335,7 @@
 
 Bool SendSetEncodings()
 {
-  char buf[sz_rfbSetEncodingsMsg + MAX_ENCODINGS * 4];
+  char buf[sz_rfbSetEncodingsMsg + MAX_ENCODINGS * sizeof(CARD32)];
   rfbSetEncodingsMsg *se = (rfbSetEncodingsMsg *)buf;
   CARD32 *encs = (CARD32 *)(&buf[sz_rfbSetEncodingsMsg]);
   int len = 0;
